addons/sale/static/src/js/product_configurator_widget.js

```diff
@@ -0,0 +1,212 @@
+     *    {integer} ev.data.changes.product_template_id : product_template ID
+     *
+     * @private
+     */
+    _onTemplateChange: function (ev) {
+
+    },
+
+    /**
+     * Hook for product_product based configurators
+     * (event, rental, ...).
+     * Should return
+     *    true if product has been configured through wizard or
+     *        the result of the super call for other wizard extensions
+     *    false if the product wasn't configurable through the wizard
+     *
+     * @param {integer} product_id
+     * @param {String} dataPointID
+     * @returns {Promise<Boolean>} stopPropagation true if a suitable configurator has been found.
+     *
+     * @private
+     */
+    _onProductChange: function (product_id, dataPointID) {
+        return Promise.resolve(false);
+    },
+
+    /**
+     * Hook for editing a configured line.
+     * The button triggering this function is only shown in Edit mode,
+     * when _isConfigurableProduct is True.
+     *
+     * @private
+     */
+    _onEditProductConfiguration: function () {
+
+    },
+
+    /**
+     * Utilities for recordData conversion
+     */
+
+    /**
+     * Will convert the values contained in the recordData parameter to
+     * a list of '4' operations that can be passed as a 'default_' parameter.
+     *
+     * @param {Object} recordData
+     *
+     * @private
+     */
+    _convertFromMany2Many: function (recordData) {
+        if (recordData) {
+            var convertedValues = [];
+            _.each(recordData.res_ids, function (res_id) {
+                convertedValues.push([4, parseInt(res_id)]);
+            });
+
+            return convertedValues;
+        }
+
+        return null;
+    },
+
+    /**
+     * Will convert the values contained in the recordData parameter to
+     * a list of '0' or '4' operations (based on wether the record is already persisted or not)
+     * that can be passed as a 'default_' parameter.
+     *
+     * @param {Object} recordData
+     *
+     * @private
+     */
+    _convertFromOne2Many: function (recordData) {
+        if (recordData) {
+            var convertedValues = [];
+            _.each(recordData.res_ids, function (res_id) {
+                if (isNaN(res_id)){
+                    _.each(recordData.data, function (record) {
+                        if (record.ref === res_id) {
+                            convertedValues.push([0, 0, {
+                                attribute_value_id: record.data.attribute_value_id.data.id,
+                                custom_value: record.data.custom_value
+                            }]);
+                        }
+                    });
+                } else {
+                    convertedValues.push([4, res_id]);
+                }
+            });
+
+            return convertedValues;
+        }
+
+        return null;
+    }
+});
+
+FieldsRegistry.add('product_configurator', ProductConfiguratorWidget);
+
+return ProductConfiguratorWidget;
+
+});
```

Comment:
Typically, we group functions into sections:
 1) (default) lifecycle functions (e.g. init willStart start destroy)
 2) public functions
 3) private functions (prefixed by _)
 4) (private) handlers (prefixed by _on)

See the [JS coding guidelines](https://github.com/odoo/odoo/wiki/Javascript-coding-guidelines)